### PR TITLE
LibWeb: Introduce DocumentState and update SessionHistoryEntry

### PR DESCRIPTION
**Pull request by kalenikaliaksandr**
I would like to start incrementally merging parts of navigables refactor https://github.com/SerenityOS/serenity/pull/18219 and this is the first chunk of this work.

These commits are from Andreas's branch where he started working on navigables. I just rebased them. I am not really feeling comfortable with PR that consists 99% of someone else's work but all my further changes will be based on Andreas's branch so I need help with getting them merged first :)

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 1]
```diff
@@ -209,15 +209,15 @@ JS::NonnullGCPtr<BrowsingContext> BrowsingContext::create_a_new_browsing_context
...
-    browsing_context->m_session_history.append(HTML::SessionHistoryEntry {
-        .url = AK::URL("about:blank"),
-        .document = document.ptr(),
-        .serialized_state = {},
-        .policy_container = {},
-        .scroll_restoration_mode = {},
-        .browsing_context_name = {},
-        .original_source_browsing_context = {},
-    });
+    JS::NonnullGCPtr<SessionHistoryEntry> new_entry = *browsing_context->heap().allocate_without_realm<SessionHistoryEntry>();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:212]: `auto`

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 2]
```diff
@@ -1009,15 +1009,15 @@ WebIDL::ExceptionOr<void> BrowsingContext::navigate_to_a_fragment(AK::URL const&
...
-    m_session_history.append(SessionHistoryEntry {
-        .url = url,
-        .document = current_entry().document,
-        .serialized_state = {},
-        .policy_container = current_entry().policy_container,
-        .scroll_restoration_mode = current_entry().scroll_restoration_mode,
-        .browsing_context_name = {},
-        .original_source_browsing_context = {},
-    });
+    JS::NonnullGCPtr<SessionHistoryEntry> new_entry = *heap().allocate_without_realm<SessionHistoryEntry>();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:1012]: `auto`

File: Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp [hunk 3]
```diff
@@ -1033,7 +1033,7 @@ WebIDL::ExceptionOr<void> BrowsingContext::navigate_to_a_fragment(AK::URL const&
 // https://html.spec.whatwg.org/multipage/browsing-the-web.html#traverse-the-history
 WebIDL::ExceptionOr<void> BrowsingContext::traverse_the_history(size_t entry_index, HistoryHandlingBehavior history_handling, bool explicit_history_navigation)
 {
-    auto* entry = &m_session_history[entry_index];
+    JS::GCPtr<SessionHistoryEntry> entry = m_session_history[entry_index];
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/BrowsingContext.cpp:1036]: `auto` (I don't think we need to turn it into a nullable GCPtr?)

File: Userland/Libraries/LibWeb/HTML/SessionHistoryEntry.h [hunk 4]
```diff
@@ -26,7 +27,12 @@ enum class ScrollRestorationMode {
 };
 
 // https://html.spec.whatwg.org/multipage/history.html#session-history-entry
-struct SessionHistoryEntry {
+struct SessionHistoryEntry final : public JS::Cell {
+    JS_CELL(SessionHistoryEntry, JS::Cell);
+
+    void visit_edges(Cell::Visitor&) override;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/SessionHistoryEntry.h:33]: `struct` members are public by default, so having a separate public section below is confusing. Arguably this is complex enough to be a class now.

File: Userland/Libraries/LibWeb/HTML/DocumentState.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-2
+class DocumentState final : public JS::Cell {
+    JS_CELL(DocumentState, JS::Cell);
+
+public:
+    virtual ~DocumentState();
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-document
+    JS::GCPtr<DOM::Document> document;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/DocumentState.h:26]: Class with public data members that are not exposed via getters is very much against https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#names:

> Data members in C++ classes should be private. [...] data members should be prefixed by "m_".

**Review by linusg — CHANGES_REQUESTED** [review]

**Comment by kalenikaliaksandr** [comment]
@linusg thank you for taking a look! I made requested changes.

File: Userland/Libraries/LibWeb/CMakeLists.txt [hunk 6]
```diff
@@ -220,6 +220,7 @@ set(SOURCES
     HTML/CustomElements/CustomElementRegistry.cpp
     HTML/DOMParser.cpp
     HTML/DOMStringMap.cpp
+    HTML/DocumentState.cpp
```
Comment (linusg) [inline Userland/Libraries/LibWeb/CMakeLists.txt:223]: This should go before `DOMParser`

File: Userland/Libraries/LibWeb/Forward.h [hunk 7]
```diff
@@ -251,6 +251,7 @@ class CustomElementDefinition;
 class CustomElementRegistry;
 class DOMParser;
 class DOMStringMap;
+class DocumentState;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/Forward.h:254]: Likewise

File: Userland/Libraries/LibWeb/HTML/DocumentState.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    // FIXME: https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-resource
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-reload-pending
+    bool m_reload_pending { false };
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-ever-populated
+    bool m_ever_populated { false };
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state-nav-target-name
+    String m_navigable_target_name { String::from_utf8_short_string(""sv) };
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/DocumentState.h:95]: This can just be a default constructed string

```
String m_navigable_target_name;
```

**Review by linusg — COMMENTED** [review]
A few more things I missed :)

**Pull request merged**